File: tightfitsudoku/submissions/accepted/TFSudokuGodmar.java
/**
 * Solution to GNY 2016/Tight-Fit Sudoku
 *
 * @author godmar
 */
import java.util.*;
import java.util.stream.*;

@SuppressWarnings("unchecked")
public class TFSudokuGodmar
{
    static char [][][] B = new char [6][6][];
    static Set<Character> ALL() {
        return Stream.of('1','2','3','4','5','6','7','8','9').collect(Collectors.toSet());
    }
    static Set<Character> [] inrow;   // choices left available to place in row
    static Set<Character> [] incol;   // choices left available to place in col
    static Set<Character> [] insqr;   // choices left available to place in square/block

    static {
        inrow = new Set[6];
        incol = new Set[6];
        insqr = new Set[6];
        for (int i = 0; i < 6; i++) {
            inrow[i] = ALL();
            incol[i] = ALL();
            insqr[i] = ALL();
        }
    }

    public static void main(String []av) {
        Scanner s = new Scanner(System.in);
        int dashes = 0;
        for (int r = 0; r < 6; r++) {
            for (int c = 0; c < 6; c++) {
                String w = s.next();
                if (w.contains("/"))
                    B[r][c] = new char[] { w.charAt(0), w.charAt(2) };
                else
                    B[r][c] = new char[] { w.charAt(0) };

                for (char v : B[r][c]) {
                    if (v != '-') {
                        inrow[r].remove(v);
                        incol[c].remove(v);
                        insqr[r/2*2+c/3].remove(v);
                    } else {
                        dashes++;
                    }
                }
            }
        }

        sudoku(dashes);
    }

    static boolean sudoku(int dashes) {
        if (dashes == 0) {
            outputBoard();
            return true;
        }

        int bestr = -1;
        int bestc = -1;
        int besti = -1;
        Set<Character> bestchoices = null;

        /* Find the r, c, i location for which the smallest number of choices exists. */
        for (int r = 0; r < 6; r++)
            for (int c = 0; c < 6; c++)
                for (int i = 0; i < B[r][c].length; i++)
                    if (B[r][c][i] == '-') {
                        Set<Character> left = new HashSet<>(inrow[r]);
                        left.retainAll(incol[c]);
                        left.retainAll(insqr[r/2*2+c/3]);

                        if (bestchoices == null || left.size() < bestchoices.size()) {
                            bestr = r;
                            bestc = c;
                            besti = i;
                            bestchoices = left;
                        }
                    }

        /* Now try all options to the (r, c, i) locations with the smallest number
         * of choices.  If the solution so far has led to a contradiction, this could
         * be none.  If an entry is now determined, it will be 1.
         */
        for (char k : bestchoices) {
            char [] cell = B[bestr][bestc];
            cell[besti] = k;

            // check x/y x < y condition
            if (cell.length == 2 && cell[1-besti] != '-' && cell[0] > cell[1]) {
                cell[besti] = '-';
                continue;
            }

            inrow[bestr].remove(k);
            incol[bestc].remove(k);
            insqr[bestr/2*2+bestc/3].remove(k);
            if (sudoku(dashes-1))
                return true;

            // undo
            cell[besti] = '-';
            inrow[bestr].add(k);
            incol[bestc].add(k);
            insqr[bestr/2*2+bestc/3].add(k);
        }
        return false;
    }

    static void outputBoard() {
        for (int r = 0; r < 6; r++) {
            for (int c = 0; c < 6; c++) {
                StringJoiner sj = new StringJoiner("/");
                for (char v : B[r][c])
                    sj.add(v+"");
                System.out.print(sj.toString() + " ");
            }
            System.out.println();
        }
    }
}

File: tightfitsudoku/submissions/accepted/tfsudoku-godmar.py


B = [] 
for i in range(6):
    B.append(map(lambda x : x.replace("/", " ").split(), raw_input().split()))

inrow = [set(map(str, range(1, 10))) for _ in range(6)]
incol = [set(map(str, range(1, 10))) for _ in range(6)]
insqr = [set(map(str, range(1, 10))) for _ in range(6)]

def ri(r, c):
    return r/2*2+c/3

left = dict()

dashes = 0
for r, row in enumerate(B):
    for c, cell in enumerate(row):    # 6 cells
        for v in cell:
            if v != '-':
                inrow[r].remove(v)
                incol[c].remove(v)
                insqr[ri(r,c)].remove(v)
            else:
                dashes += 1

choices = {}
for r, row in enumerate(B):
    for c, cell in enumerate(row):    # 6 cells
        for i, v in enumerate(cell):
            if v == '-':
                choices[(r, c, i)] = inrow[r] & incol[c] & insqr[ri(r,c)]
                
def outputboard(): # output board
    for row in B:
        for cell in row:
            if len(cell) == 2:
                print "%s/%s" % (cell[0], cell[1]),
            else:
                print cell[0],
        print

def backtrack(dashes):
    if dashes == 0:
        outputboard()
        return True

    picks = []
    for (r, c, i), ch in choices.iteritems():
        if B[r][c][i] == '-':
            picks.append((len(ch), r, c, i, ch))

    picks.sort()

    l, r, c, i, left = picks[0]
    for k in left:
        cell = B[r][c]
        cell[i] = k

        # check x/y x < y condition
        if len(cell) == 2 and cell[1-i] != '-' and cell[0] > cell[1]:
            cell[i] = '-'
            continue

        undolog = []
        for (r0, c0, i0), ch in choices.iteritems():
            if B[r0][c0][i0] == '-' and (r0 == r or c0 == c or ri(r, c) == ri(r0, c0)):
                if k in choices[(r0, c0, i0)]:
                    undolog.append((r0, c0, i0))

        for u in undolog:
            choices[u].discard(k)

        if backtrack(dashes-1):
            return True

        cell[i] = '-'
        for u in undolog:
            choices[u].add(k)

    return False

backtrack(dashes)

